keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,993 @@
+	// subset to be operational.
+	MinMember int32
+}
+```
+
+In order to associate a pod with its target PodGroup, we need to add a field call `PodGroup` in the podSpec to declare which PodGroup and Subset the pod belongs to.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+  ....
+  PodGroup PodGroupSpec
+	....
+}
+
+// - pod.spec.podGroup.name = foo
+// - pod.spec.podGroup.subset = (bar/baz)
+type PodGroupSpec struct {
+  // Name is the name of PodGroup.
+  Name   string
+  // Subset is the name of PodGroup.
+  Subset string
+}
+```
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators (like ClusterAutoscaler) can rely on it to make pending PodGroup
+schedulable. For this, we need to add a new PodGroup Controller in
+controller-manager.
+
+```go
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Represents the latest available observations of 
+	// a pod group's current state.
+	Conditions []PodGroupCondition
+
+	// Subsets is a list of SubsetStatus which represents 
+	// the current state of a single subset.
+	Subsets []SubsetStatus
+}
+
+// PodGroupCondition describes the state of a pod group 
+// at a certain point.
+type PodGroupCondition struct {
+	// Type of deployment condition.
+	Type PodGroupConditionType
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	
+	// The reason for the condition's last transition.
+	Reason string
+	
+	// A human readable message indicating details about 
+	// the transition.
+	Message string
+}
+
+type PodGroupConditionType string
+
+const (
+	// PodGroupScheduled represents status of the scheduling 
+	// process for this pod group.
+	PodGroupScheduled PodGroupConditionType = "PodGroupScheduled"
+)
+
+// SubsetStatus represents the current state of a subset.
+type SubsetStatus struct {
+	Name string
+
+	// Total number of non-terminated pods targeted by this subset. (
+	// their labels match the selector)
+	Total int32
+
+	// ScheduledAvailable is the number of pods in the subset that can be scheduled
+	// successfully, but the scheduling fails because minMember cannot be met.
+	ScheduledAvailable int32
+
+	// ScheduledAvailable is the number of pods in the subset that can't be 
+	// scheduled successfully.
+	ScheduledUnavailable int32
+
+	// Running is the number of pods targeted by this subset
+	// with a Running phase.
+	Running int32
+
+	// Succeeded is the number of pods targeted by this subset
+	// with a Succeeded phase.
+	Succeeded int32
+
+	// Failed is the number of pods targeted by this subset
+	// with a Failed phase.
+	Failed int32
```

Comment:
I don't think we should have this. The information more naturally belong to the specific workload resource (Job, Deployment, etc).

Reply:
Yeah. You are right. It depends different workload. Deleted Failed and Succeeded.